### ISSUE TITLE
Subquery condition moves to post outer join

### ISSUE THREAD

**Issue by tsdlovell** [cloudera/ibis#781]
Essentially becoming an inner join

```
import ibis
HOST = <HOST>

client = ibis.impala.connect(HOST)
db = client.database()
schema = ibis.schema([('id', 'int32'), ('desc', 'string')])
for tablename in ('left', 'right'):
    db.create_table(tablename, schema=schema, format='parquet')

left = db['left']
right = db['right']

left = left[left.id < 2]
right = right[right.id < 3]

joined = left.left_join(right, ['id', 'desc'])
joined = joined[
    [left[name].name('left_' + name) for name in left.columns] +
    [right[name].name('right_' + name) for name in right.columns]
]

print(ibis.impala.compile(joined))
```

prints
```
SELECT t0.`id` AS `left_id`, t0.`desc` AS `left_desc`, t1.`id` AS `right_id`,
       t1.`desc` AS `right_desc`
FROM default.`left` t0
  LEFT OUTER JOIN default.`right` t1
    ON t0.`id` = t1.`id` AND
       t0.`desc` = t1.`desc`
WHERE t0.`id` < 2 AND
      t1.`id` < 3
```

Workaround: use identity projection before outer join
```
left = left[left.id < 2][left]
right = right[right.id < 3][right]
```

prints
```
SELECT t0.`id` AS `left_id`, t0.`desc` AS `left_desc`, t1.`id` AS `right_id`,
       t1.`desc` AS `right_desc`
FROM (
  SELECT *
  FROM default.`left`
  WHERE `id` < 2
) t0
  LEFT OUTER JOIN (
    SELECT *
    FROM default.`right`
    WHERE `id` < 3
  ) t1
    ON t0.`id` = t1.`id` AND
       t0.`desc` = t1.`desc`
```

**Comment by wesm** [comment]
Thanks for finding this. I'll investigate and sort out a fix for 0.6.1

**Comment by wesm** [comment]
hey Dan -- sorry it took me a while to fix, but this is now all set
